components/kyma-environment-broker/internal/environmentscleanup/service.go

```diff
@@ -0,0 +1,177 @@
+		LabelSelector:   labelSelector,
+	}
+}
+
+func (s *Service) PerformCleanup() error {
+	var result *multierror.Error
+	shootsToDelete, err := s.getShootsToDelete(s.LabelSelector)
+	if err != nil {
+		log.Error(errors.Wrap(err, "while getting shoots to delete"))
+		result = multierror.Append(result, err)
+	}
+
+	var instancesDetails []instanceDetailsDTO
+	for _, shoot := range shootsToDelete {
+		runtimeId, ok := shoot.Annotations[shootAnnotationRuntimeId]
+		if !ok {
+			err = errors.New(fmt.Sprintf("shoot %q has no runtime-id annotation", shoot.Name))
+			log.Error(err)
+			result = multierror.Append(result, err)
+			continue
+		}
+		instancesDetails = append(instancesDetails, instanceDetailsDTO{
+			RuntimeID:        runtimeId,
+			CloudProfileName: shoot.Spec.CloudProfileName,
+		})
+	}
+	log.Infof("Instances Details to process: %v", instancesDetails)
+
+	instancesToDelete, err := s.getInstanceIds(instancesDetails)
+	if err != nil {
+		log.Error(errors.Wrap(err, "while getting instance IDs for runtimes"))
+		result = multierror.Append(result, err)
+	}
+
+	for _, payload := range instancesToDelete {
+		log.Infof("Triggering environment deprovisioning for instance ID %q", payload.InstanceID)
+		currentErr := s.triggerEnvironmentDeprovisioning(payload)
+		if currentErr != nil {
+			log.Error(errors.Wrapf(currentErr, "while triggering deprovisioning for instance ID %q", payload.InstanceID))
+			result = multierror.Append(result, currentErr)
+		}
+	}
+	if result != nil {
+		result.ErrorFormat = func(i []error) string {
+			var s []string
+			for _, v := range i {
+				s = append(s, v.Error())
+			}
+			return strings.Join(s, ", ")
+		}
+	}
+	return result.ErrorOrNil()
+}
+
+func (s *Service) getShootsToDelete(labelSelector string) ([]v1beta1.Shoot, error) {
+	opts := v1.ListOptions{
+		LabelSelector: labelSelector,
+	}
+	shootList, err := s.gardenerService.List(opts)
+	if err != nil {
+		return []v1beta1.Shoot{}, errors.Wrap(err, "while listing Gardener shoots")
+	}
+
+	var shoots []v1beta1.Shoot
+	for _, shoot := range shootList.Items {
+		shootCreationTimestamp := shoot.GetCreationTimestamp()
+		shootAge := time.Since(shootCreationTimestamp.Time)
+
+		if shootAge.Hours() >= s.MaxShootAge.Hours() {
+			log.Infof("Shoot %q is older than %f hours with age: %f hours", shoot.Name, s.MaxShootAge.Hours(), shootAge.Hours())
+			shoots = append(shoots, shoot)
+		}
+	}
+
+	return shoots, nil
+}
+
+type instanceDetailsDTO struct {
+	InstanceID       string
+	RuntimeID        string
+	CloudProfileName string
+}
+
+func (s *Service) getInstanceIds(instancesToDelete []instanceDetailsDTO) ([]instanceDetailsDTO, error) {
+	var runtimeIdList []string
+
+	for _, instanceDetails := range instancesToDelete {
+		runtimeIdList = append(runtimeIdList, instanceDetails.RuntimeID)
+	}
+
+	instances, err := s.instanceStorage.FindAllInstancesForRuntimes(runtimeIdList)
+	if err != nil {
+		return []instanceDetailsDTO{}, err
+	}
+
+	var toDelete []instanceDetailsDTO
```

Comment:
IMO it's an unnecessary complication, you have a list of runtimeIDs, you can just return this list of fetched instances, inside those instances you have a planID which is necessary to make deprovision operation.